cabal-lint/README.md

```diff
@@ -0,0 +1,14 @@
+The IOG Consensus team uses this `cabal-lint` tool to ensures some invariants among its `.cabal` files.
```

Comment:
I see the following behavior, currently.

```
$ cabal-lint $(find ouroboros-consensus* -name '*.cabal') | sed 's/;/\n     /g'
[E-MissingOptions] Component lib:ouroboros-consensus-protocol-test in package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal is missing required options in its `ghc-options' fields: -Wall -Wcompat -Widentities -Wincomplete-record-updates -Wincomplete-uni-patterns -Wmissing-export-lists -Wpartial-fields -Wredundant-constraints
[E-RepeatDeps] Component lib in package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal repeats dependencies in its `build-depends' fields: cardano-protocol-tpraos:lib
[E-MissingOptions] Component exe:unsoundswitch in package ouroboros-consensus/docs/report/unsoundswitch/unsoundswitch.cabal is missing required options in its `ghc-options' fields: -Wcompat -Widentities -Wincomplete-record-updates -Wincomplete-uni-patterns -Wmissing-export-lists -Wpartial-fields -Wredundant-constraints
[I-ConsistentVersions] Dependency QuickCheck:lib is consistently unconstrained
[I-ConsistentVersions] Dependency Win32:lib is consistently constrained to >=2.6.1.0
[I-ConsistentVersions] Dependency aeson:lib is consistently unconstrained
[I-ConsistentVersions] Dependency base-deriving-via:lib is consistently unconstrained
[I-ConsistentVersions] Dependency bifunctors:lib is consistently unconstrained
[I-ConsistentVersions] Dependency bimap:lib is consistently constrained to >=0.4 && <0.5
[I-ConsistentVersions] Dependency binary-search:lib is consistently unconstrained
[I-ConsistentVersions] Dependency byron-spec-chain:lib is consistently unconstrained
[I-ConsistentVersions] Dependency byron-spec-ledger:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-binary:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-crypto:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-crypto-class:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-crypto-praos:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-crypto-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-crypto-tests:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-crypto-wrapper:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-data:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-alonzo:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-alonzo-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-babbage:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-babbage-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-byron:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-byron-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-core:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-shelley:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-shelley-ma:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-shelley-ma-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-ledger-shelley-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-prelude:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-protocol-tpraos:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-slotting:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cardano-strict-containers:lib is consistently unconstrained
[I-ConsistentVersions] Dependency contra-tracer:lib is consistently unconstrained
[I-ConsistentVersions] Dependency cryptonite:lib is consistently constrained to >=0.25 && <0.31
[I-ConsistentVersions] Dependency data-default-class:lib is consistently unconstrained
[I-ConsistentVersions] Dependency deepseq:lib is consistently unconstrained
[I-ConsistentVersions] Dependency digest:lib is consistently unconstrained
[I-ConsistentVersions] Dependency fgl:lib is consistently unconstrained
[I-ConsistentVersions] Dependency file-embed:lib is consistently unconstrained
[I-ConsistentVersions] Dependency filelock:lib is consistently unconstrained
[I-ConsistentVersions] Dependency formatting:lib is consistently constrained to >=6.3 && <7.2
[I-ConsistentVersions] Dependency generic-random:lib is consistently unconstrained
[I-ConsistentVersions] Dependency generics-sop:lib is consistently unconstrained
[I-ConsistentVersions] Dependency graphviz:lib is consistently unconstrained
[I-ConsistentVersions] Dependency hashable:lib is consistently unconstrained
[I-ConsistentVersions] Dependency hedgehog-quickcheck:lib is consistently unconstrained
[I-ConsistentVersions] Dependency io-sim:lib is consistently unconstrained
[I-ConsistentVersions] Dependency logfloat:lib is consistently unconstrained
[I-ConsistentVersions] Dependency measures:lib is consistently unconstrained
[I-ConsistentVersions] Dependency microlens:lib is consistently unconstrained
[I-ConsistentVersions] Dependency optparse-applicative:lib is consistently unconstrained
[I-ConsistentVersions] Dependency optparse-generic:lib is consistently unconstrained
[I-ConsistentVersions] Dependency orphans-deriving-via:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-byron:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-byron-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-byronspec:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-cardano:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-cardano-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-cardano-tools:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-diffusion:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-mock:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-mock-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-protocol:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-protocol:lib:ouroboros-consensus-protocol-test is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-shelley:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-shelley-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-consensus-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-network-mock:lib is consistently unconstrained
[I-ConsistentVersions] Dependency ouroboros-network-protocols:lib:testlib is consistently unconstrained
[I-ConsistentVersions] Dependency pretty-show:lib is consistently unconstrained
[I-ConsistentVersions] Dependency psqueues:lib is consistently constrained to >=0.2.3 && <0.3
[I-ConsistentVersions] Dependency random:lib is consistently unconstrained
[I-ConsistentVersions] Dependency resourcet:lib is consistently unconstrained
[I-ConsistentVersions] Dependency semialign:lib is consistently constrained to >=1.1
[I-ConsistentVersions] Dependency small-steps:lib is consistently unconstrained
[I-ConsistentVersions] Dependency small-steps-test:lib is consistently unconstrained
[I-ConsistentVersions] Dependency statistics:lib is consistently unconstrained
[I-ConsistentVersions] Dependency streaming:lib is consistently unconstrained
[I-ConsistentVersions] Dependency strict-stm:lib is consistently constrained to ^>=0.2
[I-ConsistentVersions] Dependency tasty:lib is consistently unconstrained
[I-ConsistentVersions] Dependency tasty-golden:lib is consistently unconstrained
[I-ConsistentVersions] Dependency tasty-hunit:lib is consistently unconstrained
[I-ConsistentVersions] Dependency tasty-quickcheck:lib is consistently unconstrained
[I-ConsistentVersions] Dependency template-haskell:lib is consistently unconstrained
[I-ConsistentVersions] Dependency temporary:lib is consistently unconstrained
[I-ConsistentVersions] Dependency these:lib is consistently constrained to >=1.1 && <1.2
[I-ConsistentVersions] Dependency time:lib is consistently unconstrained
[I-ConsistentVersions] Dependency transformers:lib is consistently unconstrained
[I-ConsistentVersions] Dependency transformers-except:lib is consistently unconstrained
[I-ConsistentVersions] Dependency tree-diff:lib is consistently unconstrained
[I-ConsistentVersions] Dependency typed-protocols:lib is consistently unconstrained
[I-ConsistentVersions] Dependency unix:lib is consistently unconstrained
[I-ConsistentVersions] Dependency unix-bytestring:lib is consistently unconstrained
[I-ConsistentVersions] Dependency utf8-string:lib is consistently unconstrained
[I-ConsistentVersions] Dependency vector-map:lib is consistently unconstrained
[E-InconsistentVersions] Dependency base:lib is unconstrained in component test:test of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-analyser of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-synthesizer of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component test:test of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component test:test of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component lib:ouroboros-consensus-protocol-test of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component test:test of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-infra of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to ^>=4.14.1.0 in component exe:unsoundswitch of package ouroboros-consensus/docs/report/unsoundswitch/unsoundswitch.cabal
      but constrained to >=4.9 && <4.17 in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component lib of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency base16-bytestring:lib is unconstrained in component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
      but constrained to >=1.0 in component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal
[E-InconsistentVersions] Dependency binary:lib is unconstrained in component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.8 && <0.11 in component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency bytestring:lib is unconstrained in component test:test of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-synthesizer of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component test:test of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component lib:ouroboros-consensus-protocol-test of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component test:test of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.10 && <0.12 in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component lib of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency cborg:lib is unconstrained in component test:test of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component test:test of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component test:test of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.2.2 && <0.3 in component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency containers:lib is unconstrained in component test:test of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component test:test of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component test:test of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.5 && <0.7 in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency directory:lib is unconstrained in component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-synthesizer of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=1.3 && <1.4 in component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency filepath:lib is unconstrained in component test:test of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-synthesizer of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component test:test of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=1.4 && <1.5 in component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency io-classes:lib is unconstrained in component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to ^>=0.3 in component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency mtl:lib is unconstrained in component test:test of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=2.2 && <2.3 in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency nothunks:lib is unconstrained in component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.1.2 && <0.2 in component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency ouroboros-consensus:lib is unconstrained in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component test:test of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-analyser of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-synthesizer of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component test:test of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component test:test of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-infra of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to ^>=0.1 in component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal
[E-InconsistentVersions] Dependency ouroboros-network:lib is unconstrained in component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to ^>=0.2 in component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal
[E-InconsistentVersions] Dependency ouroboros-network-api:lib is unconstrained in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
      but constrained to ^>=0.1 in component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal
[E-InconsistentVersions] Dependency ouroboros-network-framework:lib is unconstrained in component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to ^>=0.2 in component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal
[E-InconsistentVersions] Dependency ouroboros-network-protocols:lib is unconstrained in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component test:test of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component test:test of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to ^>=0.2 in component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency quickcheck-state-machine:lib is unconstrained in component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.7.0 in component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
[E-InconsistentVersions] Dependency quiet:lib is unconstrained in component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.2 && <0.3 in component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency serialise:lib is unconstrained in component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component test:test of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.2 && <0.3 in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component lib of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency sop-core:lib is unconstrained in component lib of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.5 && <0.6 in component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency text:lib is unconstrained in component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=1.2 && <1.3 in component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentVersions] Dependency vector:lib is unconstrained in component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal
      but constrained to >=0.12 && <0.13 in component lib of package ouroboros-consensus/ouroboros-consensus.cabal
[E-InconsistentLanguages] The default language is unset in component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
      but set to Haskell2010 in component lib of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component test:test of package ouroboros-consensus-byron-test/ouroboros-consensus-byron-test.cabal, component lib of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component exe:db-converter of package ouroboros-consensus-byron/ouroboros-consensus-byron.cabal, component lib of package ouroboros-consensus-byronspec/ouroboros-consensus-byronspec.cabal, component lib of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component test:test of package ouroboros-consensus-cardano-test/ouroboros-consensus-cardano-test.cabal, component lib of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-analyser of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component exe:db-synthesizer of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component test:test of package ouroboros-consensus-cardano-tools/ouroboros-consensus-cardano-tools.cabal, component lib of package ouroboros-consensus-cardano/ouroboros-consensus-cardano.cabal, component lib of package ouroboros-consensus-diffusion/ouroboros-consensus-diffusion.cabal, component lib of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component test:test of package ouroboros-consensus-mock-test/ouroboros-consensus-mock-test.cabal, component lib of package ouroboros-consensus-mock/ouroboros-consensus-mock.cabal, component lib of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component lib:ouroboros-consensus-protocol-test of package ouroboros-consensus-protocol/ouroboros-consensus-protocol.cabal, component lib of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component test:test of package ouroboros-consensus-shelley-test/ouroboros-consensus-shelley-test.cabal, component lib of package ouroboros-consensus-shelley/ouroboros-consensus-shelley.cabal, component lib of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-consensus of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-infra of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component test:test-storage of package ouroboros-consensus-test/ouroboros-consensus-test.cabal, component exe:unsoundswitch of package ouroboros-consensus/docs/report/unsoundswitch/unsoundswitch.cabal, component lib of package ouroboros-consensus/ouroboros-consensus.cabal
$ echo $?
1
```

Reply:
That output raises one main question: do we want to introduce a new invariant, requiring that every dependency lists non-empty bounds?